File: AUGUST-2025/21-08-2025/p1504.cpp
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int rows = mat.size();
        int cols = mat[0].size();
        int sum = 0;

        for (int i = 0; i < rows; i++){
            for (int j = 1; j < cols; j++){
                if (mat[i][j] == 0) continue;
                else mat[i][j] += mat[i][j-1];
            }
        }

        for (int r = 0; r < rows; r++){
            for (int c = 0; c < cols; c++){
                if (mat[r][c] == 0) continue;
                int width = INT_MAX;
                for (int dr = r; dr < rows; dr++){
                    width = min(width, mat[dr][c]);
                    if (width == 0) break;
                    sum += width;
                }
            }
        }

        return sum;
    }
};

File: AUGUST-2025/21-08-2025/p1504.java
class Solution {
    public int numSubmat(int[][] mat) {
        int rows = mat.length;
        int cols = mat[0].length;

        int[] heights = new int[cols];
        int total = 0;

        for (int i = 0; i < rows; i++) {
            // Update the heights array in-place for current row
            for (int j = 0; j < cols; j++) {
                heights[j] = mat[i][j] == 0 ? 0 : heights[j] + 1;
            }

            // Pass the same updated height to the histogram method
            total += countFromHistogram(heights);
        }

        return total;
    }

    private int countFromHistogram(int[] height) {
        int len = height.length;
        Stack<Integer> stack = new Stack<>();
        int[] dp = new int[len];
        int res = 0;

        for (int j = 0; j < len; j++) {
            while (!stack.isEmpty() && height[stack.peek()] >= height[j]) {
                stack.pop();
            }

            if (stack.isEmpty()) {
                dp[j] = height[j] * (j + 1);
            } else {
                int prevIndex = stack.peek();
                dp[j] = dp[prevIndex] + height[j] * (j - prevIndex);
            }

            res += dp[j];
            stack.push(j);
        }

        return res;
    }
}